feature_tests/steps/common.py

```diff
@@ -0,0 +1,91 @@
+from uuid import UUID
+
+import parse
+from behave import register_type
+
+
+@parse.with_pattern(
+    r"\{[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}\}"
+)
+def parse_uuid(value):
+    return UUID(value)
+
+
+def parse_string(value):
+    return str(value)
+
+
+@parse.with_pattern(r"-?[0-9]+")
+def parse_number(value):
+    return int(value)
+
+
+@parse.with_pattern(r"[0-9A-Z]{1,5}")
+def parse_ods_code(value):
+    return value
+
+
+@parse.with_pattern(r"(a|an)")
+def parse_a_or_an(value):
+    return value
+
+
+@parse.with_pattern(r"[a-z0-9_]+(\.[a-z0-9_]+)*")
+def parse_path(value):
+    return value
+
+
+# Register the different datatypes used in the step definitions.
+register_type(
+    UUID=parse_uuid,
+    String=parse_string,
+    Number=parse_number,
+    OdsCode=parse_ods_code,
+    An=parse_a_or_an,
+    Path=parse_path,
+)
+
+
+def catch_errors(func):
+    """
+    Decorator that will catch an exception and record it on the context
+    """
+
+    def inner(context, *args, **kwargs):
+        try:
+            func(context, *args, **kwargs)
+        except Exception as e:
+            context.error = e
+
+    return inner
+
+
+def assert_type_matches(obj, expected):
+    actual = type(obj).__name__
+    assert actual == expected, "Type mismatch: expected {expected} got {actual}"
+
+
+def parse_value(v):
+    """
+    Use the parsing methods above to turn a value into a proper type.
+    """
+
+    def __parse_value(v, *funcs):
+        [head, *tail] = funcs
+        try:
+            return head(v)
+        except:
+            return __parse_value(v, *tail)
+
+    return __parse_value(v, parse_uuid, parse_number, parse_ods_code, parse_string)
+
+
+def read_value_from_path(obj, path) -> any:
```

Comment:
i understand what this is doing, but that's only because i understand low level python (there are cases and object types that this definitely won't work for since __dict__ doesn't necessarily return attributes). At a minimum needs a docstring to explain that it's attempting to retrieve a value by json path in a python object's attributes